### ISSUE TITLE
Spring Boot Starter HATEOAS should register the Jackson2HalModule by Default 

### ISSUE THREAD

**Issue by jzampieron** [spring-projects/spring-boot#6962]
Problem:
In a µservices setup trying to consume REST repositories from Spring Data Rest JPA using Feign, Ribbon and Hystrix the default object mapper on the client side doesn't understand the "_embedded" field.
 
Proposed Enhancement: 
The default `Jackson2ObjectMapperBuilderCustomizer` for spring boot starter HATEOAS should register the module `Jackson2HalModule` by default.

This makes it substantially more "works out of the box" to consume Spring Data Rest exported Repositories using a Spring Boot application. 

Only applicable to Spring Boot 1.4.0 and later.

**Comment by wilkinsona** [comment]
@jzampieron I'm not familiar with how a Feign-based client uses Jackson and gets configured with its `ObjectMapper`. Can you provide a small example that illustrates the problem?

> Only applicable to Spring Boot 1.4.0 and later.

Why do you believe this is only applicable to 1.4 and later?

**Comment by jzampieron** [comment]
There's no bug here. Its a DX (developer experience) enhancement request. 

I'm hardly a Feign expert, but AFAIK, it uses the default `ObjectMapper` from the `HttpMessageConverter` processing chain to deserialize the JSON response body into an object.

In order to get Feign to properly deserialize the "_embedded" elements (and others) from the spring-data-jpa-rest server, I had to add the following bean:

```
@Component
public class SpringDataRestJacksonCustomizer
   implements Jackson2ObjectMapperBuilderCustomizer
{

   /* (non-Javadoc)
    * @see org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer#customize(org.springframework.http.converter.json.Jackson2ObjectMapperBuilder)
    */
   @Override
   public void customize( Jackson2ObjectMapperBuilder jacksonObjectMapperBuilder )
   {
      jacksonObjectMapperBuilder
         .modulesToInstall( new Jackson2HalModule() );
   }

}
```

The overall experience is suboptimal in versions of Spring boot prior to v1.4.0, but the `Jackson2ObjectMapperBuilderCustomizer` only exists as of v1.4.0, so requesting the above enhancement in this form is only directly applicable to v1.4.0. 

Versions prior would require alternate approaches.

I think the overall DX can be improved by automatically installing the Jackson2HalModule by default if Spring Boot HATEOAS is on the classpath.

**Comment by wilkinsona** [comment]
Thanks for the clarification. I think that applying the `Jackson2HalModule` to the default `ObjectMapper` would be too invasive. Given its use across the entire application, there's no guarantee that the `ObjectMapper` is only being used to (de)serialize HAL-format JSON.

The auto-configuration for Spring HATEOAS switches on Spring HATEOAS's `Jackson2HalModule` configuration but it's very careful to not touch things that it shouldn't. For example, it adds an HTTP message converter for `application/hal+json` with a custom `ObjectMapper` configured with the module, leaving the converter for `application/json` alone.

In short, I don't think we should do anything in Spring Boot as applying it to the default `ObjectMapper` is too broad and we don't know anything about Feign so we can't be more precise. It may be possible for Spring Cloud Netfix to do something that's specific to Feign-based clients so you might want to raise a [Spring Cloud Netflix](https://github.com/spring-cloud/spring-cloud-netflix) issue.
